shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java

```diff
@@ -360,5 +381,69 @@ public void dispatchSemanticsAction(
     public int systemGestureInsetRight = 0;
     public int systemGestureInsetBottom = 0;
     public int systemGestureInsetLeft = 0;
+    public List<DisplayFeature> displayFeatures = List.of();
+  }
+
+  /**
+   * Area of the viewport obstructed by a feature, eg. a hinge, a fold crease or camera cutout
+   *
+   * <p>Similar to {@link androidx.window.DisplayFeature}, used for reporting Fold / Hinge areas.
+   * The reason we define our own model is that we also want to support cutouts.
+   */
+  public static final class DisplayFeature {
+    public final Rect bounds;
+    public final DisplayFeatureType type;
+    public final int state;
+
+    public DisplayFeature(Rect bounds, DisplayFeatureType type, int state) {
+      this.bounds = bounds;
+      this.type = type;
+      this.state = state;
+    }
+
+    public DisplayFeature(Rect bounds, DisplayFeatureType type) {
+      this.bounds = bounds;
+      this.type = type;
+      this.state = 0; // UNKNOWN
+    }
+  }
+
+  /**
+   * Types of display features that can obstruct the viewport.
+   *
+   * <p>Some, like FOLD, can be reported without actually impeding drawing on the screen. They are
+   * useful for knowing where the display is bent or has a crease. The {@link DisplayFeature} bounds
+   * can be 0-width in such cases.
+   */
+  public enum DisplayFeatureType {
+    /**
+     * We do not know this type of display feature yet. This can happen if WindowManager is updated
+     * with new types.
+     */
+    UNKNOWN(0),
```

Comment:
Or this could happen when the window is not yet attached when setWindowManagerDisplayFeatures is called via onLayout right? 

I wonder if we should delay the sending of the very first viewport metrics until a window is attached. Worth testing whether during initialization, we get a quick succession of { window size X*Y }, then very quickly after, { window size X*Y, hinge at rect Z }, causing the Flutter application listening to the mediaquery to rapidly build 2 different layouts (wasting the first one).

Reply:
Will test and see. If there is a delay, I don't think it's large enough to warrant both calls.